File: src/features/typing/data/java/valid_parentheses.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Map;

class Solution {
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        Map<Character, Character> pairs = Map.of(')', '(', ']', '[', '}', '{');
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') {
                stack.push(c);
                continue;
            }
            Character top = stack.poll();
            if (top == null || !top.equals(pairs.get(c))) {
                return false;
            }
        }
        return stack.isEmpty();
    }
}

File: src/features/typing/data/dart/valid_parentheses.dart
class Solution {
  bool isValid(String s) {
    final stack = <String>[];
    const pairs = {')': '(', ']': '[', '}': '{'};
    for (int i = 0; i < s.length; i++) {
      final c = s[i];
      if (c == '(' || c == '[' || c == '{') {
        stack.add(c);
        continue;
      }
      if (stack.isEmpty || stack.removeLast() != pairs[c]) {
        return false;
      }
    }
    return stack.isEmpty;
  }
}
